components/script/dom/textcontrol.rs

```diff
@@ -13,43 +18,53 @@ use dom::node::{Node, NodeDamage, window_from_node};
 use script_traits::ScriptToConstellationChan;
 use textinput::{SelectionDirection, SelectionState, TextInput};
 
-pub trait TextControl: DerivedFrom<EventTarget> + DerivedFrom<Node> {
-    fn textinput(&self) -> &DomRefCell<TextInput<ScriptToConstellationChan>>;
+pub trait TextControlElement: DerivedFrom<EventTarget> + DerivedFrom<Node> {
     fn selection_api_applies(&self) -> bool;
     fn has_selectable_text(&self) -> bool;
     fn set_dirty_value_flag(&self, value: bool);
+}
+
+pub struct TextControlSelection<'a, E: TextControlElement> {
+    element: &'a E,
```

Comment:
So the problem you're facing here is this line. Since the SM GC doesn't trace the Rust stack, it does not know that there is an actively borrowed reference to the HTML{Input, TextArea}Element, and as a result it may collect it while Rust still has a borrowed reference to it, making it "unsafe".

The solution here is to create a pointer type that SM recognizes and can trace, and there are 2 of them that we can use here: either `Dom<T>` or `DomRoot<T>`. I _think_ `DomRoot<T>` may make more sense here, since all of the uses of the `TextControlSelection` struct does not bind to any identifier and is treated as a temporary.

Reply:
I'm not sure that replacing the type with something else is valuable. Our borrowed pointers to DOM types are fine, because they can only be obtained from properly rooted values. I'm inclined to leave the `allow` line as it is.

Reply:
That is also fine, since I don't think it's actually possible to have the GC collect the element at all in any case.

Reply:
I actually tried changing it to use a `DomRoot<T>`, which took ages to figure out and once I finally got there I still had the `unrooted_must_root` error... hence I reverted back to what I've submitted here and asked for advice!